src/Features/Core/Portable/BraceCompletion/AbstractBraceCompletionService.cs

```diff
@@ -0,0 +1,178 @@
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            return OpeningBrace == brace && !syntaxFactsService.IsInNonUserCode(tree, openingPosition, cancellationToken);
+        }
+
+        ///<inheritdoc cref="IBraceCompletionService.IsInsideCompletedBracesAsync(int, Document, CancellationToken)"/>
+        public async Task<BraceCompletionContext?> IsInsideCompletedBracesAsync(int caretLocation, Document document, CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var leftToken = root.FindTokenOnLeftOfPosition(caretLocation);
+            var rightToken = root.FindTokenOnRightOfPosition(caretLocation);
+
+            if (IsValidOpeningBraceToken(leftToken) && IsValidClosingBraceToken(rightToken))
+            {
+                return new BraceCompletionContext(document, leftToken.GetLocation().SourceSpan.Start, rightToken.GetLocation().SourceSpan.End, caretLocation);
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Checks if the already inserted token is a valid opening token at the position in the document.
+        /// </summary>
+        protected virtual Task<bool> IsValidOpenBraceTokenAtPositionAsync(SyntaxToken token, int position, Document document, CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            if (!IsParentSkippedTokensTrivia(syntaxFactsService, token))
+            {
+                return SpecializedTasks.False;
+            }
+
+            return Task.FromResult(IsValidOpeningBraceToken(token) && token.SpanStart == position);
+        }
+
+        /// <summary>
+        /// Checks that the current position is a valid location.
+        /// Used to determine if overtype should be allowed.
+        /// </summary>
+        protected static async Task<bool> CheckCurrentPositionAsync(Document document, int? currentPosition, CancellationToken cancellationToken)
+        {
+            // make sure auto closing is called from a valid position
+            if (currentPosition == null)
+            {
+                return false;
+            }
+
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            return !syntaxFactsService.IsInNonUserCode(tree, currentPosition.Value, cancellationToken);
+        }
+
+        protected static bool IsParentSkippedTokensTrivia(ISyntaxFactsService syntaxFactsService, SyntaxToken token)
+            => token.Parent != null && !syntaxFactsService.IsSkippedTokensTrivia(token.Parent);
```

Comment:
i don't see how this possibly works.  the method is called IsParentSkippedTokensTrivia, but then you return ` !syntaxFactsService.IsSkippedTokensTrivia(token.Parent);`  

this also means we're missing a test afaict.

Reply:
the usages of the check are correct, I just renamed this wrong - I'm going to switch the impl to be the name as it makes more sense to me to do the check that way